File: 781-rabbits-in-forest/781-rabbits-in-forest.cpp
const int N = 1e3 + 10;
int hsh[N];

class Solution {
public:
    int numRabbits(vector<int>& answers) {
        for (int i = 0; i < N; ++i) hsh[i] = 0;
        int ans = 0;
        for (auto &ele : answers)
            hsh[ele]++;
        ans += (hsh[0]);
        for (int i = 1; i < N; ++i) {
            if (hsh[i]) {
                int q = hsh[i] / (i + 1);
                int r = hsh[i] % (i + 1);
                if (r == 0) ans += (i * q + q);
                else ans += (q) ? (i * q + i + 1 + q) : (i + 1);
            }
        }
        return ans;
    }
};

File: 781-rabbits-in-forest/781-rabbits-in-forest.java
class Solution {
    public static int formula(int n, int count) {
        int tempAns = 0;
        int q = count / (n + 1);
        int r = count % (n + 1);
        tempAns = tempAns + (q * (n + 1));
        if (r > 0) tempAns += (n + 1);
        return tempAns;
    }

    public int numRabbits(int[] a) {
        int N = 1000 + 10;
        int[] hsh = new int[N];
        for (int ele : a)
            hsh[ele]++;
        int ans = 0;
        ans += hsh[0];
        for (int i = 1; i < hsh.length; ++i) {
            if (hsh[i] != 0)
                ans += formula(i, hsh[i]);
        }
        return ans;
    }
}